Cstr verify safety contracts of unsafe functions strlen, from_bytes_with_nul_unchecked

rajathkotyal opened the pull request:
Towards #150

Annotates and verifies the safety contracts for the following unsafe functions:

`from_bytes_with_nul_uncheked` `core::ffi::c_str`
`strlen` `core::ffi::c_str`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

rajathkotyal commented on library/core/src/ffi/c_str.rs:
```diff
@@ -875,4 +875,187 @@ mod verify {
+            assert!(c_str.is_safe());
+        }
+    }
+
```
Relevant code below this

Yenyun035 commented on library/core/src/ffi/c_str.rs:
```diff
@@ -875,4 +875,187 @@ mod verify {
+    fn check_strlen() {
+        const MAX_SIZE: usize = 32;
+        let mut string: [u8; MAX_SIZE] = kani::any();
+        let nul_position = kani::any::<usize>() % MAX_SIZE;
```
How about `kani::any_where(|x| *x >= 0 && x < MAX_SIZE)`?

rajathkotyal commented on library/core/src/ffi/c_str.rs:
```diff
@@ -875,4 +880,82 @@ mod verify {
     }
-}
\ No newline at end of file
+
```
The stubbing part is a bit unclear : 

Yeah. Since Kani doesnt support external runtime functions, I tried stubbing them, but the functions are in a nested manner (the extern "C" block) , we cannot stub nested functions since its not in the scope and is invoked during runtime. We could try to explicitly mock this function but Im not sure if that is expected.

i.e. the extern "C" block is taken out of this function maintaining functionality of the code OR, we mock the working of the strlen. Reference : https://model-checking.github.io/kani/rfc/rfcs/0002-function-stubbing.html
```
#[inline]
#[unstable(feature = "cstr_internals", issue = "none")]
#[cfg_attr(bootstrap, rustc_const_stable(feature = "const_cstr_from_ptr", since = "1.81.0"))]
#[rustc_allow_const_fn_unstable(const_eval_select)]
const unsafe fn strlen(ptr: *const c_char) -> usize {
    const fn strlen_ct(s: *const c_char) -> usize {
        let mut len = 0;

        // SAFETY: Outer caller has provided a pointer to a valid C string.
        while unsafe { *s.add(len) } != 0 {
            len += 1;
        }

        len
    }

    #[inline]
    fn strlen_rt(s: *const c_char) -> usize {
        extern "C" {
            /// Provided by libc or compiler_builtins.
            fn strlen(s: *const c_char) -> usize;
        }

        // SAFETY: Outer caller has provided a pointer to a valid C string.
        unsafe { strlen(s) }
    }

    intrinsics::const_eval_select((ptr,), strlen_ct, strlen_rt)
}
```

Yenyun035 replied:
@rajathkotyal FYI, strlen was updated:
```rust
const unsafe fn strlen(ptr: *const c_char) -> usize {
    const_eval_select!(
        @capture { s: *const c_char = ptr } -> usize:
        if const {
            let mut len = 0;

            // SAFETY: Outer caller has provided a pointer to a valid C string.
            while unsafe { *s.add(len) } != 0 {
                len += 1;
            }

            len
        } else {
            extern "C" {
                /// Provided by libc or compiler_builtins.
                fn strlen(s: *const c_char) -> usize;
            }

            // SAFETY: Outer caller has provided a pointer to a valid C string.
            unsafe { strlen(s) }
        }
    )
}
```

rajathkotyal replied:
Thank you, this resolves the nested module blocker

rajathkotyal commented on library/core/src/ffi/c_str.rs:
```diff
@@ -882,6 +887,84 @@ mod verify {
+
+    // Stub implementation for strlen: https://model-checking.github.io/kani/rfc/rfcs/0002-function-stubbing.html
+    #[cfg(kani)]
+    unsafe fn stub_strlen(s: *const c_char) -> usize {
```
can we stub only strlen_rt here?

rajathkotyal commented on library/core/src/ffi/c_str.rs:
```diff
@@ -939,4 +990,75 @@ mod verify {
+    // Proof harness that uses the const-evaluated length
+    // #[kani::proof_for_contract(strlen)]
+    #[kani::proof]
+    #[kani::unwind(32)]
```
Since strlen isnt public, I wasnt able to access it in the verify module. Figured of verifying using the function below. It works but Im not sure if this is expected.

celinval replied:
Where did you place the verify module? You should be able to wrap this function with the verify module. A private function is still available for all descendants of the module where it is defined.

What error were you getting?

rajathkotyal replied:
My bad, needed to invoke using the super keyword. Its working now. Thanks!

rajathkotyal commented on library/core/src/ffi/c_str.rs:
```diff
@@ -939,4 +990,75 @@ mod verify {
+    // Stub for external C strlen
+    #[allow(dead_code)]
+    mod stubs {
+        use super::*;
```
Based on my research on stubs : 
https://github.com/model-checking/kani/issues/2423
https://github.com/firecracker-microvm/firecracker/blob/dbd9a84b11a63b5e5bf201e244fe83f0bc76792a/src/rate_limiter/src/lib.rs#L607-L639
https://github.com/model-checking/kani/pull/2658

I figured we can stub the external strlen function individually and write a proof for it, But this seems to not work as expected and gives me this error.

```
   Compiling dummy v0.1.0 (/Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/target/kani_verify_std)
error: Failed to find contract closure `__kani_recursion_check_strlen` in function `ffi::c_str::strlen`

error: could not compile `core` (lib) due to 1 previous error
error: Failed to execute cargo (exit status: 101). Found 1 compilation errors.
```

celinval replied:
The problem here is that the strlen you are stubbing is the one you are trying to check.

There's no way for Kani to stub a nested function. Can you pass `-Z c-ffi` to Kani and remove the stub to see if that would work? Otherwise, we can add support for strlen to Kani. I believe CBMC already has a model available.

rajathkotyal replied:
Hi @celinval , thank you, using the flag seems to work, I have made some changes to the code and pushed it. Pl review

rajathkotyal commented on library/core/src/ffi/c_str.rs:
```diff
@@ -939,4 +990,37 @@ mod verify {
+        let mut string: [u8; MAX_SIZE] = kani::any();
+        let nul_position: usize = kani::any_where(|x| *x < MAX_SIZE);
+        string[nul_position] = 0;
+        
```
We are using proof_for_contract and wouldn't want this assumption of checking nul before last byte be in our harness, But added here to improve brevity. Can move the pre condition check to the function definition requires clause as well. Something like : 

```
#[requires(!ptr.is_null() && {
    let mut found_null = false;
    let mut valid = true;
    for i in 0..isize::MAX as usize {
        if unsafe { *ptr.add(i) } == 0 {
            found_null = true;
            break;
        } else {
            // Ensure no interior null bytes
            if unsafe { *ptr.add(i) } == 0 {
                valid = false;
                break;
            }
        }
    }
    found_null && valid
})]
```

rajathkotyal replied:
Update :  I tried the above method of including the precondition in the requires clause and received this error : 
```

VERIFICATION RESULT:
 ** 1 of 242 failed
Failed Checks: This is a placeholder message; Kani doesn't support message formatted at runtime
 File: "/Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/library/core/src/panicking.rs", line 418, in panicking::assert_failed_inner

VERIFICATION:- FAILED
Verification Time: 8.302092s

Summary:
Verification failed for - ffi::c_str::verify::check_strlen_contract
Complete - 0 successfully verified harnesses, 1 failures, 1 total.
```

tautschnig replied:
Can you run Kani with the additional options `--harness check_strlen_contract --output-format old --enable-unstable --no-reachability-checks --cbmc-args --trace` to find the source line of the assertion that is failing?

rajathkotyal replied:
I tried this command :

``` 
./scripts/run-kani.sh --kani-args --enable-unstable --harness ffi::c_str::verify --output-format old --no-assertion-reach-checks --cbmc-args --trace
```
 
but cbmc seems to be interfering with "--enable-unstable" flag and giving me  
``` 
Unknown option: --enable-unstable
Summary:
Verification failed for - ffi::c_str::verify::check_strlen_contract
Complete - 0 successfully verified harnesses, 1 failures, 1 total.
```

Not sure if my command format is wrong

rajathkotyal replied:
Checked the logs, It was a `ptr` dereferencing error where since we weren't checking for interior nul bytes before calling the `strlen` func, it was panicking since it encountered nul bytes. Figured we need to keep the assumption before passing to the function as well as in the requires clause. Pushed a new commit now, it is working.

rajathkotyal replied:
FYI : These were the error I encountered  : 

```
Violated property:
  file /Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/library/core/src/panicking.rs function panicking::assert_failed_inner line 418 thread 0
  This is a placeholder message; Kani doesn't support message formatted at runtime
  FALSE

[panicking::assert_failed_inner.assertion.2] line 418 This is a placeholder message; Kani doesn't support message formatted at runtime: FAILURE

/Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/library/core/src/panicking.rs function panicking::panic_nounwind
```

celinval commented on library/core/src/ffi/c_str.rs:
```diff
@@ -755,6 +760,39 @@ impl AsRef<CStr> for CStr {
+// - ptr must point to a valid null-terminated string
+// - null terminator must be within isize::MAX bytes from ptr
+#[requires(!ptr.is_null() && {
+    let mut found_null = false;
+    let mut valid = true;
+    for i in 0..isize::MAX as usize {
+        if unsafe { *ptr.add(i) } == 0 {
+            if found_null {
+                // Already found a null before, so this is an extra null
+                valid = false;
+                break;
+            }
+            found_null = true;
+            break;
+        }
+    }
+    found_null && valid
```
What about something like this?
```rust
let mut next = ptr;
loop {
  if !can_dereference(next) {
    break false;
  } else if unsafe { *next == 0 } {
    break true;
  }
  next = next.wrapping_add(1);
}
```

rajathkotyal replied:
Thanks. Just updated the code.

celinval commented on library/core/src/ffi/c_str.rs:
```diff
@@ -755,6 +760,39 @@ impl AsRef<CStr> for CStr {
+// Postconditions:
+// - Returns length before null terminator
+// - Length must be less than isize::MAX
+#[ensures(|&result| result < isize::MAX as usize && {
+    let mut valid = true;
+    for i in 0..result {
+        if unsafe { *ptr.add(i) } == 0 {
+            valid = false;
+            break;
+        }
+    }
+    valid && unsafe { *ptr.add(result) } == 0
+})]
```
I think you can simplify this a bit since we mostly care about safety. You could use `[slice_from_raw_parts](https://doc.rust-lang.org/std/ptr/fn.slice_from_raw_parts.html)` to create a slice ptr, check that we can dereference it, and ensure the last byte is 0.

rajathkotyal replied:
Sure. Updated the code.

tautschnig commented on library/core/src/ffi/c_str.rs:
```diff
@@ -942,6 +992,39 @@ mod verify {
+        string[nul_position] = 0;
+        
+        // Ensure no null bytes before nul_position
+        for i in 0..nul_position {
+            kani::assume(string[i] != 0);
+        }
+        
+        let ptr = string.as_ptr() as *const c_char;
+        
+        unsafe {
+            let len = super::strlen(ptr);
+            
+            // Length must be less than isize::MAX
+            assert!(len < isize::MAX as usize);
+            // Length must equal nul_position
+            assert_eq!(len, nul_position);
+            // no null in between
+            for i in 0..len {
+                assert!(*ptr.add(i) != 0);
+            }
```
In the precondition you use `contains` while here you use (both for the assumption and the assertion) explicit loops. I am wondering whether there is a performance difference between those approaches? Would you mind experimenting with 1) consistently using `contains` and 2) consistently using loops?

rajathkotyal replied:
postcondition you mean?

tautschnig replied:
Actually true of both: you also have it in a precondition in line 439.

rajathkotyal replied:
Oh gotcha. That was for `from_bytes_with_nul_unchecked` . Will check and update

tautschnig replied:
Thank you for gathering this data! I'd suggest you use a uniform solution, be it `contains` or loops. I'll leave it to you to pick.

celinval commented on library/core/src/ffi/c_str.rs:
```diff
@@ -755,6 +762,30 @@ impl AsRef<CStr> for CStr {
+// Preconditions:
+// - ptr must be non-null and properly aligned
+// - ptr must point to a valid null-terminated string within isize::MAX bytes
+#[requires(!ptr.is_null() && {
```
Once your other PR merges, can you please update this to use the new function?

rajathkotyal replied:
Sure will do

rajathkotyal commented on library/core/src/ffi/c_str.rs:
```diff
@@ -755,6 +762,30 @@ impl AsRef<CStr> for CStr {
+// - Length must be less than isize::MAX
+// - No null bytes before returned length
+#[ensures(|&result| result < isize::MAX as usize && unsafe {
+    let slice = core::slice::from_raw_parts(ptr, result + 1);
+    // Check no nulls before result and null at result
+    !slice[..result].contains(&0) && slice[result] == 0
```
Do you mean we remove this entirely or either keep this or the assertions in the code? Im confused how we are going to validate the results if we remove the ensures caluse

celinval replied:
Since we've been looking at safety verification at this point, we need to ensure no UB occurs and that all objects that may be modified by the function are left in a safe state.

Further checks are nice to have but not necessary. If we do add functional post-conditions, I would prefer a simpler ensures clause so it's not hard to maintain

rajathkotyal replied:
Simplified code by retaining essential checks and removed redundant verifications from the proof harness, pl review @celinval

rajathkotyal commented on library/core/src/ffi/c_str.rs:
```diff
@@ -942,6 +992,39 @@ mod verify {
+            }
+            // additional checks
+            assert_eq!(*ptr.add(len), 0);
+            assert!(len <= MAX_SIZE - 1);
```
Okay. Just a quick check, do you mean to say this code below is a sufficient amount of check for strlen? @celinval 

conditions : 
```
#[requires(!ptr.is_null() && is_null_terminated(ptr))]
#[ensures(|&result| result < isize::MAX as usize)]
```

proof  harness :  
```
    #[kani::proof_for_contract(super::strlen)]
    #[kani::unwind(33)]
    fn check_strlen_contract() {
        const MAX_SIZE: usize = 32;
        let string: [u8; MAX_SIZE] = kani::any();
        
        let ptr = string.as_ptr() as *const c_char;
        
        unsafe {
            let result_len = super::strlen(ptr);
            assert_eq!(string[result_len], 0);
        }
    }
```

Note : this does successfully pass.
```
Checking harness ffi::c_str::verify::check_strlen_contract...

VERIFICATION RESULT:
 ** 0 of 117 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 5.4615207s

Complete - 1 successfully verified harnesses, 0 failures, 1 total.
```

Yenyun035 replied:
I think this one doesn't check for all string <= max_size but only the string == max_size.

rajathkotyal commented:
Hi @celinval , incorporated all changes and removed redundant checks, could you review once again and approve the workflow if everything checks out? Let me know if there is anything missing. Thanks!

Yenyun035 commented:
The workflow failed likely due to the following error. I ran on my Macbook Pro (Apple M1 Pro, 16 GB RAM, Sequoia 15.0.1)
```
Checking harness ffi::c_str::verify::check_strlen_contract...

VERIFICATION RESULT:
 ** 1 of 158 failed (157 undetermined)
Failed Checks: call to foreign "C" function `strlen` is not currently supported by Kani. Please post your example at https://github.com/model-checking/kani/issues/2423
 File: "/Users/yew005/Docs/Academic/CMU/Fall24/practicum/vrs/library/core/src/ffi/c_str.rs", line 798, in ffi::c_str::strlen::{closure#2}::{closure#1}::runtime

VERIFICATION:- FAILED
** WARNING: A Rust construct that is not currently supported by Kani was found to be reachable. Check the results for more details.
Verification Time: 1.4835527s

Summary:
Verification failed for - ffi::c_str::verify::check_strlen_contract
Complete - 0 successfully verified harnesses, 1 failures, 1 total.
```

celinval commented:
> The workflow failed likely due to the following error. I ran on my Macbook Pro (Apple M1 Pro, 16 GB RAM, Sequoia 15.0.1)
> 
> ```
> Checking harness ffi::c_str::verify::check_strlen_contract...
> 
> VERIFICATION RESULT:
>  ** 1 of 158 failed (157 undetermined)
> Failed Checks: call to foreign "C" function `strlen` is not currently supported by Kani. Please post your example at https://github.com/model-checking/kani/issues/2423
>  File: "/Users/yew005/Docs/Academic/CMU/Fall24/practicum/vrs/library/core/src/ffi/c_str.rs", line 798, in ffi::c_str::strlen::{closure#2}::{closure#1}::runtime
> 
> VERIFICATION:- FAILED
> ** WARNING: A Rust construct that is not currently supported by Kani was found to be reachable. Check the results for more details.
> Verification Time: 1.4835527s
> 
> Summary:
> Verification failed for - ffi::c_str::verify::check_strlen_contract
> Complete - 0 successfully verified harnesses, 1 failures, 1 total.
> ```

That means we need to enable C-FFI support like #204

Yenyun035 commented:
> That means we need to enable C-FFI support like https://github.com/model-checking/verify-rust-std/pull/204
@celinval Got it. Thank you for mentioning that. This means #204 needs to be merged first.

carolynzech commented:
@Yenyun035 can you fix the failing Rust tests and then we'll take a look if there's still issues?

Yenyun035 commented:
@carolynzech Sure. The failed Rust tests were due to unused `is_null_terminated` helper function we created for verifying `strlen`. Is it better to annotate it with `#[cfg(kani)]` or `allow_unused_import`?

Yenyun035 commented:
@carolynzech Sure. I annotated `is_null_terminated` helper function with `#[cfg(kani)]` now. I think it should fix the failed rust tests. If not, I'll try to figure out another solution.

carolynzech commented on library/core/src/ffi/c_str.rs:
```diff
@@ -11,6 +11,8 @@ use crate::slice::memchr;
 use crate::{fmt, ops, slice, str};
 
 use crate::ub_checks::Invariant;
+use crate::ub_checks::can_dereference;
```
```suggestion
#[allow(unused_imports)]
use crate::ub_checks::can_dereference;
```

Yenyun035 replied:
Thank you.

Yenyun035 commented:
@tautschnig Fixed. Thank you.

carolynzech reviewed (approved):
Nice!

rajathkotyal commented:
Thanks @Yenyun035 !
